### ISSUE TITLE
Enable thinking models

### ISSUE THREAD

**Issue by rlouf** [outlines-dev/outlines#1627]
A v0 would be allowing users to instantiate the logits processor `ThinkingLogitsProcessor` with the end of thinking token and pass it to the generator.

```python
model = ...
logits_processor = ThinkingLogitsProcessor("</think>", output_type)
generator = Generator(model, processor=logits_processor)
```

The implementation is simple:

1. While `</think>` has not been observed, `__call__` is a pass-through.
2. After `<think>` has been generated, mask the logits to generate the type passed by the user.

**Comment by TheooJ** [comment]
Agree that this would be very useful. For now, the best I've found is to add ``reasoning: str`` to the top of the type, which significantly improves performance (both for thinking and non-thinking models).

Could there also be a ``start_processing_logits=</think>`` directly in the Generator ? In any case I like the proposed API too.

**Comment by RobinPicard** [comment]
Adding `reasoning: str` to the top of the type is when you're using a Pydantic class to represent a JSON schema, right?

**Comment by rlouf** [comment]
A quick note, unrelated to the discussion. During internal evaluations, we've come across a couple of instances where the model "thinks" beyond the total context length. This makes me think we could add an option to limit the number of thinking tokens.
